File: cpp/producer-consumer/producer-consumer-test/ConsumerTest.cpp
#include <vector>

#include <gtest/gtest.h>

#include <ChunkData.h>
#include <DataQueue.h>
#include <Consumer.h>

using namespace producer_consumer;
using std::vector;
using std::shared_ptr;

class ConsumerTest : public testing::Test
{
protected:
	ConsumerTest()
		: consumer(&queue)
	{
	}

	virtual void SetUp()
	{
	}

	DataQueue queue;

	Consumer consumer;
};

TEST_F(ConsumerTest, HandlePrimeNumber)
{
	queue.Inqueue(shared_ptr<ChunkData>(new ChunkData({ 1, 3, 7 })));

	ASSERT_TRUE(consumer.Consume());
	ASSERT_EQ(0, queue.Size());
}

TEST_F(ConsumerTest, HandleBoundaryZero)
{
	queue.Inqueue(shared_ptr<ChunkData>(new ChunkData({ 0, 0, 0 })));

	Consumer consumer(&queue);
	EXPECT_FALSE(consumer.Consume());
	EXPECT_EQ(0, queue.Size());
}

File: java/producer-consumer/src/test/java/producer_consumer/ConsumerTest.java
package producer_consumer;

import static org.hamcrest.CoreMatchers.is;
import static org.junit.Assert.assertThat;

import org.junit.After;
import org.junit.AfterClass;
import org.junit.Before;
import org.junit.BeforeClass;
import org.junit.Test;

public class ConsumerTest {

    private static Consumer consumer;

    private static DataQueue queue;

    @BeforeClass
    public static void globalSetUp() {
        ConsumerTest.queue = new DataQueue();
        ConsumerTest.consumer = new Consumer(queue);
    }

    @AfterClass
    public static void globalTearDown() {
        // TODO put class level tear down here
    }

    @Before
    public void setUp() {
        // TODO put method level set up here
    }

    @After
    public void tearDown() {
        // TODO put method level tear down here
    }

    @Test
    public void handlePrimeNumber() {
        queue.inqueue(new ChunkData(new int[] { 1, 3, 7 }));
        assertThat(consumer.consume(), is(true));
    }

    @Test
    public void HandleBoundaryZero() {
        queue.inqueue(new ChunkData(new int[] { 0, 0, 0 }));
        assertThat(consumer.consume(), is(false));
    }

}
